### DISCUSSION TITLE
Distinguish tab-close from temporary disconnect (blocked thread, lid, network)

### DISCUSSION THREAD

**Discussion by rodja** [zauberzeug/nicegui#2410] [Ideas / Feature Requests]
As we discussed in #2286, for truly non-blocking execution in JavaScript we should sepearate the connection to the server from the main execution thread. A simple example to reproduce the problem is

```py
label = ui.label()
ui.timer(0.1, lambda: label.set_text(f'{time.time():.3f}'))
ui.button('Print', on_click=lambda: ui.run_javascript('print()'))
```

The time does not get updated anymore as long as the print dialog is opened. That leads to a connection loss when `reconnect_timeout` is reached.

**Comment by rodja** [comment]
I guess #2073 suffered from this. And maybe #2715, too.

**Reply by BaccanoMob** [reply]
Also #4881 (js script to prevent page refesh/reload by warning in alert dialog)

**Comment by evnchn** [comment]
Got https://github.com/zauberzeug/nicegui/pull/5784

**Comment by falkoschindler** [comment]
## Stepping back — what are we actually disambiguating?

The repro in the body and the Web-Worker take in #5784 both treat this as "keep the connection alive while the main thread is blocked." But the underlying issue is broader: the server can't tell **"user closed the tab"** from **"main thread is blocked / lid is closed / network blipped"** — they all look like a Socket.IO disconnect, and we apply the same 3 s `reconnect_timeout` to all of them. Anything that disambiguates those two cases solves the dialog problem as a special case.

The Service-Worker / Web-Worker direction proves liveness by sending a beat from a thread the dialog can't freeze. PR #5784 implements that with a Web Worker POSTing to `/_nicegui/heartbeat` every ~2 s. It works, but it pays a permanent cost: **one POST per connected client every ~2 s, forever**, on every deployment, even when no one has ever blocked the main thread. At 1 000 connected clients that's ~500 RPS of HTTP traffic that's a no-op server-side 99 % of the time, doubled on top of the existing Socket.IO ping baseline.

The inverse approach: have the page **explicitly say goodbye when it leaves**, and treat anything else as "probably still around, wait longer." The expensive case (proving liveness) becomes a one-shot beacon at end-of-life; the common case (healthy connection) emits zero new HTTP traffic.

## Sketch

**Client (in `nicegui.js`, ~5 lines):**

```js
window.addEventListener("pagehide", (e) => {
  if (e.persisted) return; // bfcache — page may resume, don't signal close
  navigator.sendBeacon(
    `${options.prefix}/_nicegui/closing`,
    new Blob([JSON.stringify({ client_id: window.clientId })], { type: "application/json" }),
  );
});
```

`sendBeacon` is queued by the browser and survives unload — no async, no race with tab teardown.

**Server (`nicegui.py`, new endpoint):**

```python
@app.post('/_nicegui/closing')
async def _closing(request: Request) -> Response:
    try:
        client_id = (await request.json()).get('client_id')
    except ValueError:
        return Response(status_code=400)
    if client_id and (client := Client.instances.get(client_id)):
        client.handle_explicit_close()
    return Response(status_code=200)
```

**Server (`client.py`):**

No new config knob. `reconnect_timeout` already means exactly "how long to wait after disconnect for the browser to come back." We just **bump its default from 3 s to ~5 min**, and short-circuit the wait when we know reconnect won't happen:

- Implicit disconnect (no pagehide received) → wait `reconnect_timeout`. Same code path as today.
- Beacon arrives → `handle_explicit_close()` cancels any pending delete task and runs `self.delete()` immediately (or after a 1 s grace if we want to absorb double-clicks).

So the change is: one new endpoint, one new JS listener, one new method on `Client`, and a default-value bump. The cancel+recreate plumbing from #5784 can be dropped — there's only ever one delete task per document, and the beacon either short-circuits it or it's not.

## Behaviour vs. heartbeat PR

| Scenario                           | Heartbeat PR (#5784)                                            | Beacon proposal                                             |
| ---------------------------------- | --------------------------------------------------------------- | ----------------------------------------------------------- |
| `alert()` for 30 s, user clicks OK | Survives ✅                                                     | Survives ✅ (within bumped `reconnect_timeout`)             |
| Laptop lid closed for 10 min       | **Dies** ❌ (worker can't run while suspended)                  | Survives ✅                                                 |
| Mobile tab backgrounded for 5 min  | Likely dies (worker throttled)                                  | Survives ✅                                                 |
| User closes tab                    | 3 s cleanup ✅                                                  | Immediate cleanup ✅                                        |
| User force-quits browser           | 3 s cleanup ✅                                                  | Cleanup after `reconnect_timeout` (e.g., 5 min) — slower ⚠️ |
| Bot does `GET /` without WS        | 60 s prune ✅                                                   | 60 s prune ✅ (unchanged)                                   |
| Steady-state HTTP load             | +1 POST per client per 2 s                                      | Zero                                                        |
| New code                           | Worker file + endpoint + 2 new dicts + reschedule + JS plumbing | Endpoint + 1 new flag + 5 lines of JS + default bump        |

The two designs solve overlapping but different problems. Worth deciding which we actually have.

## Trade-off — is this worse?

The honest cost is **slower zombie cleanup** in the force-quit / OS-kill / sendBeacon-dropped case. A laptop that gets killed mid-session leaves the client around for `reconnect_timeout` (5 min default?) instead of 3 s. Bounded UI-tree memory per zombie, eventually swept. Slack / Discord / Notion all live with similar trade-offs.

Note this only affects clients that *did* connect via WebSocket. Bots / scrapers / health checks that only do `GET /` without ever opening a WS go through `Client.prune_instances` (60 s sweep) instead, which is unchanged.

In exchange we get:

- The lid-close case stops being a footgun.
- Mobile background no longer kills sessions.
- No worker, no periodic traffic, no per-deployment HTTP baseline doubled.

## What I'd want to discuss before either lands

1. **Is the heartbeat solving the right problem?** This thread originally focused on blocking dialogs, but bumping `reconnect_timeout` and adding the beacon fixes a strictly larger set of problems — including some I think users hit more often than `alert()` (lid close, brief network drops).
2. **Could we ship both?** They're not mutually exclusive — the beacon is cheap and the heartbeat is the failsafe. But if the beacon covers ~90 % of cases, the heartbeat's complexity is harder to justify.
3. **Does bumping the `reconnect_timeout` default break anyone?** It's a behaviour change — clients that today vanish in 3 s would linger for minutes. I think this is a strict improvement, but worth calling out in the changelog.

If the consensus lands on the beacon, #5784 closes in favour of that work. If the heartbeat is preferred, the comparison at least makes the trade-off explicit before we commit a watchdog to the codebase.

**Reply by evnchn** [reply]
In public discourse, I've disagreed to bumping timeout for default for a while, but forgot what's the term for it. 

I've found it, and it can be considered a "Slow DOS attack". Read more at https://en.wikipedia.org/wiki/Slow_DoS_attack

In general I think it is a bad idea, if the attacker only needs to send very little bandwidth, to occupy a big chunk of server resources. 

@falkoschindler do you still think it's a good idea? Moreover, your idea requires the reconnect_timeout to be set to N minutes for N minutes the main thread is blocked, which isn't very scalable.

**Reply by falkoschindler** [reply]
Fair pushback — Slow-DoS is the right framing for it. Let me concede where I think you're right and push back where I think it's bounded; we don't have to land on a path today.

## DoS amplification — real, but more bounded than Slowloris

Slowloris is byte-asymmetric: a few bytes of attacker traffic occupies a full server connection. The beacon-with-bumped-timeout isn't quite that shape — each zombie still costs the attacker a full page load (HTML + Vue + assets) and a WebSocket handshake. Per-connection effort, not per-byte. Amplification is "5 min / 3 s ≈ 100×", not the 1000–10000× Slowloris achieves. For comparison, `Client.prune_instances` already gives bots a 60 s window to accumulate `Client` instances by hammering `GET /` without ever opening a WS, and we've lived with that.

**Where you're right and I underweighted it:** zombie clients keep their `ui.timer`s ticking server-side. That's real ongoing CPU per zombie, not just memory. 100× the timer work for abandoned pages is a cost I didn't price in.

## On "N-minute blocks need N-minute timeouts" — structurally true, practically minor

Yes, the beacon is a step function: between "disconnect detected" and "user dismisses dialog" the page can't signal anything (main thread frozen), so there's no way to extend the deadline mid-block. Block < timeout → survives; block > timeout → dies. The heartbeat avoids this because the worker is a separate thread (modulo lid-close, where neither helps).

But how often does the threshold bite? `alert`/`confirm`/`prompt` are almost always <60 s, `print()` a few minutes worst case, heavy compute usually <30 s. The genuinely-long case is "user walked away from an open dialog" — real, but a long tail. With `reconnect_timeout` set generously, the beacon covers most realistic blocking-dialog cases; the heartbeat covers the long tail too.

So heartbeat wins on the "arbitrarily long block" axis, but for the realistic distribution a beacon with a sane timeout gets us most of the way.

---

I don't think we need to settle the path forward here — both mechanisms have legitimate cases and trade-offs. Worth more eyes on the discussion before we commit either way.

**Reply by evnchn** [reply]
@falkoschindler I think we may end up with a modular approach with different "Keep-Alive providers" which user can mix and match depending on their page logic, their security boundaries, and their user needs. How does that sound?

**Reply by falkoschindler** [reply]
That's what Claude suggested as well. But this would require two more or less complex implementations and the (unexperienced) user might not even find or understand both techniques. Though I haven't given up the hope to find a one-fits-all solution. Need to think about it.

**Reply by evnchn** [reply]
But one-fits-all solution would be hard because the _moment_ we crank default value for `reconnect_timeout` or make our defaults insecure we end up with a GHSA, so it has an unreasonably strong arm in the decision-making.

**Comment by falkoschindler** [comment]
The discussion is getting a bit hard to follow, so I asked Claude for an overview. It helped me very much to distinguish the different problems (A, B, C and D) we're trying to solve and the approaches (longer timeout, web worker, immediate prune):

## 1. Scenarios we're trying to handle

The first clarifying move: the "client disappeared" umbrella covers four very different situations, and each demands a different mechanism. Lumping them together is what makes the discussion go in circles.

| #   | Scenario                                     | Client state                                          | Why current behavior fails                                                                |
| --- | -------------------------------------------- | ----------------------------------------------------- | ----------------------------------------------------------------------------------------- |
| A   | Lid closed / device asleep                   | No signal at all, possibly for hours                  | Server prunes after reconnect timeout; user comes back to a dead page                     |
| B   | Poor / flaky network                         | Signal lost intermittently, seconds–minutes           | Same as A but shorter; reconnect logic kicks in                                           |
| C   | Main thread blocked (popup, modal, heavy JS) | Browser alive, but JS event loop can't pump heartbeat | Heartbeat misses → server thinks client is gone, even though it's there                   |
| D   | Tab/page actually closing                    | Client is truly gone                                  | Server still holds the client until reconnect timeout — wasted memory, larger DOS surface |

## 2. Mechanisms on the table

| Mechanism                                                   | Helps scenario          | DOS risk                                                        | Complexity | Notes                                                                                          |
| ----------------------------------------------------------- | ----------------------- | --------------------------------------------------------------- | ---------- | ---------------------------------------------------------------------------------------------- |
| Longer reconnect timeout                                    | A, B                    | **High** — zombies accumulate; attacker can spawn cheap clients | Trivial    | The core tradeoff falko identified — no way around it                                          |
| Web worker heartbeat (always-on)                            | C                       | Low                                                             | Medium     | Doesn't help A/B at all; "frequent POSTs" is the objection                                     |
| Web worker **watchdog** (POST only when main thread silent) | C                       | Low                                                             | Medium     | Resolves the traffic objection; main thread answers normally, worker only speaks up on silence |
| Full websocket in worker                                    | C (and indirectly more) | Low                                                             | **High**   | `window.socket` is part of user-visible API — breaking change risk                             |
| `pagehide` → immediate prune                                | D                       | **Reduces** DOS surface                                         | Low        | Independent of the others; pure win, especially if reconnect timeout grows                     |

## 3. Key observations falling out of the matrix

- **A/B and C are orthogonal.** Web workers don't help offline clients. Longer timeouts don't help busy-but-online clients. So "web worker vs. longer timeout" is a false dichotomy — they're complements.
- **The DOS tradeoff is unavoidable for A/B.** No mechanism dodges it; the only question is _where_ on the curve we sit (timeout length, optional opt-in, per-route override).
- **`pagehide` is strictly additive.** It only _reduces_ the live-client count, never increases it. It also makes a longer reconnect timeout cheaper (fewer zombies because explicit-close clients leave immediately).
- **Watchdog beats always-on heartbeat** for the worker design — it answers the "too much traffic" objection without losing the benefit.
- **Full-websocket-in-worker is the riskiest** because of the `window.socket` API surface. Worth a separate audit before committing to it.

## 4. Suggested decision order

1. Land `pagehide` pruning — uncontroversial, makes everything else cheaper.
2. Decide on web worker design (watchdog vs. full WS migration) — addresses scenario C cleanly with low DOS risk.
3. _Then_ discuss reconnect-timeout knobs for A/B with the DOS tradeoff explicit and the zombie count already reduced by step 1.

## 5. Open questions

- Is the reconnect timeout currently global, or per-client? Per-route opt-in could let apps that need long offline tolerance pay the DOS cost only where it matters.
- Does `window.socket` actually have user-visible callers in the wild, or is it internal-only in practice? Determines feasibility of full-WS-in-worker.
- Can the watchdog distinguish "main thread blocked" from "tab backgrounded by browser throttling"? If not, the worker may need to be conservative about how loudly it speaks up.
